tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -3392,3 +3393,305 @@ def call(self, inputs, state):
+      erase_vector = array_ops.expand_dims(
+        math_ops.sigmoid(erase_add_list[i * 2]), axis=1)
+      add_vector = array_ops.expand_dims(
+        math_ops.tanh(erase_add_list[i * 2 + 1]), axis=1)
+      erase_M = array_ops.ones_like(M) - math_ops.matmul(w, erase_vector)
+      M = M * erase_M + math_ops.matmul(w, add_vector)
+
+    output = math_ops.matmul(
+      array_ops.concat([controller_output] + read_vector_list, axis=1),
+      self._output_kernel)
+    output = nn_ops.bias_add(output, self._output_bias)
+    output = clip_ops.clip_by_value(output, -self.clip_value, self.clip_value)
+
+    return output, NTMControllerState(
+      controller_state=controller_state, read_vector_list=read_vector_list,
+      w_list=w_list, M=M, time=prev_state.time+1)
+
+  def _expand(self, x, dim, N):
+    return array_ops.concat(
+      [array_ops.expand_dims(x, dim) for _ in range(N)], axis=dim)
+
+  def _addressing(self, k, beta, g, s, gamma, prev_M, prev_w):
+    # Sec 3.3.1 Focusing by Content
+
+    k = array_ops.expand_dims(k, axis=2)
+    inner_product = math_ops.matmul(prev_M, k)
+    k_norm = math_ops.sqrt(math_ops.reduce_sum(
+      math_ops.square(k), axis=1, keepdims=True))
+    M_norm = math_ops.sqrt(math_ops.reduce_sum(
+      math_ops.square(prev_M), axis=2, keepdims=True))
+    norm_product = M_norm * k_norm
+    
+    # eq (6)
+    K = array_ops.squeeze(inner_product / (norm_product + 1e-8))
+
+    K_amplified = math_ops.exp(array_ops.expand_dims(beta, axis=1) * K)
+    
+    # eq (5)
+    w_c = K_amplified / math_ops.reduce_sum(K_amplified, axis=1, keepdims=True)
+
+    # Sec 3.3.2 Focusing by Location
+
+    g = array_ops.expand_dims(g, axis=1)
+    
+    # eq (7)
+    w_g = g * w_c + (1 - g) * prev_w
+
+    s = array_ops.concat([s[:, :self.shift_range + 1],
+      array_ops.zeros([s.get_shape()[0],
+        self.memory_size - (self.shift_range * 2 + 1)]),
+      s[:, -self.shift_range:]], axis=1)
+    t = array_ops.concat([array_ops.reverse(s, axis=[1]),
+      array_ops.reverse(s, axis=[1])], axis=1)
+    s_matrix = array_ops.stack(
+      [t[:, self.memory_size - i - 1:self.memory_size * 2 - i - 1]
+        for i in range(self.memory_size)],
+      axis=1)
+    
+    # eq (8)
+    w_ = math_ops.reduce_sum(
+      array_ops.expand_dims(w_g, axis=1) * s_matrix, axis=2)
+    w_sharpen = math_ops.pow(w_, array_ops.expand_dims(gamma, axis=1))
+
+    # eq (9)
+    w = w_sharpen / math_ops.reduce_sum(w_sharpen, axis=1, keepdims=True)
+
+    return w
+
+  def zero_state(self, batch_size, dtype):
+    with vs.variable_scope('init', reuse=self.reuse):
+      read_vector_list = [array_ops.zeros([batch_size, self.memory_vector_dim])
+        for _ in range(self.read_head_num)]
+
+      w_list = [array_ops.zeros([batch_size, self.memory_size])
+        for _ in range(self.read_head_num + self.write_head_num)]
+
+      controller_init_state = self.controller.zero_state(batch_size, dtype)
+
+      M = self._expand(vs.get_variable('init_M',
```

Comment:
is there any way to move all initialization to `build` and have `zero_state` return all-zeros tensors?  then in `call` you can, e.g., add `self._init_m` to state.M if `time==0` ?

Reply:
I have tried this, and for some reason which I cannot figure out - this impacts the convergence time for the model. Is it fine to leave this as is given that it affects performance (significantly) and I couldn't find a work-around.

Reply:
Do you have a PR or branch available and repro instructions -- to be able to compare convergence performance and see how you're running the test?  The reason is we're moving to TF 2.0 and variable scope is going away; we'll be doing everything "like Keras", and if we want this cell to be included we need to find a way forward that doesn't create variables when zero_state is called.

Reply:
hmm; i wonder if you could just use self.add_variable instead of vs.get_variable here.  would that work?

do you expect folks to call zero_state() more than once?  if not, i suppose you could check if it has been called more than once -- and raise a ValueError (or warn that you're reusing the same Variable), otherwise call self.add_variable.

Reply:
Is the issue creating the variable in `zero_state` or using `vs.get_variable` (or both)? I will need to re-run the tests to double check but using `vs.get_variable` in the `build` method also worked fine, so I can set M to a constant in `zero_state` and then call `vs.get_variable` in `build` if that's works for you? (will need to double check this works in terms of convergence times)

Reply:
To test convergence times for yourself, see: https://github.com/MarkPKCollier/NeuralTuringMachine `run_tasks.py`

The following commands will let you run experiments:

`python run_tasks.py --experiment_name=test --mann=ntm --task=copy --use_local_impl=False --curriculum=uniform --optimizer=Adam --learning_rate=0.001 --num_train_steps=6000 --verbose=False`

`python run_tasks.py --experiment_name=test --mann=ntm --task=associative_recall --use_local_impl=False --curriculum=uniform --optimizer=Adam --learning_rate=0.001 --num_train_steps=5000 --verbose=False --max_seq_len=6 --num_bits_per_vector=6`

Reply:
I think vs.get_variable should be replaced with self.add_variable if possible (this can be done outside of build).  If you can do it in `build` and set `M` to a constant in zero_state that would be even better.

Reply:
Ok, this is where the issue arises. When I replace vs.get_variable with self.add_variable - I get much slower convergence, looking at the docs I don't see why this would be the case but I have tested that the initializer, trainable, etc. parameters are the same in both case.

So given that for whatever reason the convergence rates are different with what should be a change that doesn't make a difference, is it acceptable to use vs.get_variable in the `build` method and avoid creating a variable in `zero_state`?